File: 2445-reachable-nodes-with-restrictions/2445-reachable-nodes-with-restrictions.java
class Solution {
    public int reachableNodes(int n, int[][] edges, int[] restricted) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : edges) {
            graph.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
            graph.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
        }
        Set<Integer> seen = new HashSet<>();
        seen.add(0);
        for (int i : restricted) {
            seen.add(i);
        }
        int res = 0;
        Queue<Integer> q = new LinkedList<>();
        q.add(0);
        while (!q.isEmpty()) {
            int v = q.poll();
            res++;
            for (int nb : graph.get(v)) {
                if (!seen.contains(nb)) {
                    seen.add(nb);
                    q.add(nb);
                }
            }
        }
        return res;
    }
}

File: 2445-reachable-nodes-with-restrictions/2445-reachable-nodes-with-restrictions.py
class Solution:
    def reachableNodes(self, n: int, edges: List[List[int]], restricted: List[int]) -> int:
        graph = defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
        seen = set(restricted)
        seen.add(0)
        q = deque([0])
        res = 0
        while q:
            v = q.popleft()
            res += 1            
            for n in graph[v]:
                if n not in seen:
                    seen.add(n)
                    q.append(n)
        return res
